chunk/chunk_store.go

```diff
@@ -491,59 +605,78 @@ func (c *AWSStore) lookupChunksForMetricName(ctx context.Context, userID string,
 
 func (c *AWSStore) lookupChunksForMatcher(ctx context.Context, userID string, bucket bucketSpec, metricName model.LabelValue, matcher *metric.LabelMatcher) (ByID, error) {
 	hashValue := hashValue(userID, bucket.bucket, metricName)
+	f := func(rangePrefix []byte) (ByID, error) {
+		input := &dynamodb.QueryInput{
+			TableName: aws.String(bucket.tableName),
+			KeyConditions: map[string]*dynamodb.Condition{
+				hashKey: {
+					AttributeValueList: []*dynamodb.AttributeValue{
+						{S: aws.String(hashValue)},
+					},
+					ComparisonOperator: aws.String("EQ"),
+				},
+				rangeKey: {
+					AttributeValueList: []*dynamodb.AttributeValue{
+						{B: rangePrefix},
+					},
+					ComparisonOperator: aws.String(dynamodb.ComparisonOperatorBeginsWith),
+				},
+			},
+			ReturnConsumedCapacity: aws.String(dynamodb.ReturnConsumedCapacityTotal),
+		}
+
+		chunkSet := ByID{}
+		var processingError error
+		var pages, totalDropped int
+		defer func() {
+			queryRequestPages.Observe(float64(pages))
+			queryDroppedMatches.Observe(float64(totalDropped))
+		}()
+		if err := c.dynamo.queryPages(ctx, input, func(resp interface{}, lastPage bool) (shouldContinue bool) {
+			var dropped int
+			dropped, processingError = processResponse(resp.(*dynamodb.QueryOutput), &chunkSet, matcher)
+			totalDropped += dropped
+			pages++
+			return processingError != nil && !lastPage
+		}); err != nil {
+			log.Errorf("Error querying DynamoDB: %v", err)
+			return nil, err
+		} else if processingError != nil {
+			log.Errorf("Error processing DynamoDB response: %v", processingError)
+			return nil, processingError
+		}
+
+		sort.Sort(ByID(chunkSet))
+		return chunkSet, nil
+	}
```

Comment:
I'd strongly consider making this a separate method/function, and just pass through all the values it's closing over.